### ISSUE TITLE
Cargo install rustfmt-nightly fails on nightly

### ISSUE THREAD

**Issue by orenbenkiki** [rust-lang/rustfmt#4776]
I'm trying to install `rustfmt-nightly` and the build fails. This tries to install `v1.4.21`. One of the dependencies is `rustc-ap-rustc_data_structures` version `v677.0.0` which fails to compile because it uses `#![feature(optin_builtin_traits)]` which has been removed. I see that the latest version of `rustc-ap-rustc_data_structure` is `v712.0.0` so perhaps just bumping its version would solve the problem?

```
$ rustc --version
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
$ rustup default nightly
$ cargo install rustfmt-nightly
...
  Installing rustfmt-nightly v1.4.21
...
  Compiling rustc-ap-rustc_data_structures v677.0.0
error[E0557]: feature has been removed
  --> /home/obk/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_data_structures-677.0.0/src/lib.rs:17:12
   |
17 | #![feature(optin_builtin_traits)]
   |            ^^^^^^^^^^^^^^^^^^^^ feature has been removed
   |
   = note: renamed to `auto_traits`
...
```

**Comment by calebcartwright** [comment]
Thanks for reaching out but going to close this.

The recommended way to add rustfmt to your system is via rustup (https://github.com/rust-lang/rustfmt#on-the-nightly-toolchain), or in some more edge-type cases one could also grab the binaries from our [releases](https://github.com/rust-lang/rustfmt/releases).

Compiling rustfmt itself is tricky, as rustfmt utilizes large chunks of the rustc internals, and those internals typically require very specific versions of nightly toolchains for compilation. This means that there's a relatively small range of nightly releases that can be used to compile any given of version of rustfmt, and the range of those supported nightly toolchains can vary from a couple days to a couple months.

We've not been distributing rustfmt through crates.io for a while, which is why the "latest" version was last published 7+ months ago and is very old relative to what's available through the supported channels (v1.4.21 vs. v1.4.36 available via rustup).

If you _really_ wanted to try to get that old version from crates, then you'd need to find and grab a suitable nightly toolchain. I don't know exactly what that range would be from ~7 months ago, but there's likely a few from around that time frame (Sept. 2020) that would technically work.

**Comment by orenbenkiki** [comment]
It isn't that I want to install it directly, but that's what `cargo make` does, specifically when running on Travis:

```
[tasks.install-rustfmt]
description = "Installs cargo rustfmt plugin."
category = "Development"
env.CFG_RELEASE = { value = "${CARGO_MAKE_RUST_VERSION}", condition = { env_not_set = ["CFG_RELEASE"] } }
env.CFG_RELEASE_CHANNEL = { value = "${CARGO_MAKE_RUST_CHANNEL}", condition = { env_not_set = ["CFG_RELEASE_CHANNEL"] } }
install_crate = { crate_name = "rustfmt-nightly", rustup_component_name = "rustfmt", binary = "rustfmt", test_arg = "--help" }
```

Ok, that's a problem with `cargo make`. I should open an issue there asking it will switch to trying `rustup component add` instead of `cargo install`, but that would only be a partial solution. Even if I work around it by pre-installing `rustfmt`, by executing `rustup component add rustfmt` before ever running `cargo make`, I get:

```
error: component 'rustfmt' for target 'x86_64-unknown-linux-gnu' is unavailable for download for channel nightly

Sometimes not all components are available in any given nightly. If you don't need the component, you can remove it with:

    rustup component remove --toolchain nightly --target x86_64-unknown-linux-gnu rustfmt
```

Which makes sense since as you said, `rustfmt` is not available for every `nightly` version. But wait... my environment is up to date and `rustfmt` works for me. What gives? Looking at my setup:

```
$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: latest update on 2021-03-30, rust version 1.53.0-nightly (48691ea6e 2021-03-29)
info: skipping nightly which is missing installed component 'rustfmt'
info: syncing channel updates for 'nightly-2021-03-29-x86_64-unknown-linux-gnu'
info: latest update on 2021-03-29, rust version 1.53.0-nightly (4a20eb6a9 2021-03-28)
info: skipping nightly which is missing installed component 'rustfmt'
info: syncing channel updates for 'nightly-2021-03-28-x86_64-unknown-linux-gnu'
info: latest update on 2021-03-28, rust version 1.53.0-nightly (9b0edb7fd 2021-03-27)
info: skipping nightly which is missing installed component 'rustfmt'
info: syncing channel updates for 'nightly-2021-03-27-x86_64-unknown-linux-gnu'
info: latest update on 2021-03-27, rust version 1.53.0-nightly (5e65467ef 2021-03-26)
info: skipping nightly which is missing installed component 'rustfmt'
info: syncing channel updates for 'nightly-2021-03-26-x86_64-unknown-linux-gnu'
info: latest update on 2021-03-26, rust version 1.53.0-nightly (52e3dffa5 2021-03-25)
info: skipping nightly which is missing installed component 'rustfmt'
info: syncing channel updates for 'nightly-2021-03-25-x86_64-unknown-linux-gnu'
info: checking for self-updates

   stable-x86_64-unknown-linux-gnu unchanged - rustc 1.51.0 (2fd73fabe 2021-03-23)
  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)

info: cleaning up downloads & tmp directories
$~/rust/total-space$ rustc --version
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
$~/rust/total-space$ cargo --version
cargo 1.52.0-nightly (90691f2bf 2021-03-16)
```

Ah - "skipping nightly which is missing installed component `rustfmt`". So `rustup` is kind enough to only update to the every-once-in-a-while version that has `rustfmt` (presumably because I have `rustfmt` installed). Hmmm... on the one hand, good it doesn't break my environment, on the other hand, I'm not getting the latest compiler...

Travis installs Rust by running ` curl -sSf https://build.travis-ci.com/files/rustup-init.sh | sh -s -- --default-toolchain=$TRAVIS_RUST_VERSION --profile=minimal -y` - in this case, `TRAVIS_RUST_VERSION=nightly`. This apparently gives it the very latest version even though it doesn't include `rustfmt`.

So, two related questions:

* Any advice on how to get `rustfmt` working on Travis on nightly builds? Right now my builds succeed/fail intermittently depending on whether `rustfmt` is available, because I use `cargo make` and one of the steps is verifying the code is properly formatted (which I do want to keep as a test of the code).

* Any way to update my local environment to the latest nightly compiler and still have a working `rustfmt`? Right now when I develop locally I don't always get the latest compiler just because I happen to have `rustfmt` installed - and being able to run `cargo fmt` is extremely useful, not something I want to give up.

On its face, both issues would be resolved if there was a way of installing `rustfmt` by downloading a statically linked precompiled binary - the latest one, even if compiled with a slightly older `rustc` version - independently of the version of the currently installed `rustc`. If that means using `curl` instead of `rustup`, so be it. Is there such an alternative?

BTW, for context - my crate needs `feature(trait_alias)` so I don't do stable builds at all. In fact I don't even have the stable toolchain installed on some of my development machines (and Travis only installs nightly when doing nightly builds, which is reasonable).

**Comment by calebcartwright** [comment]
TBH I found your last comment a bit hard to follow as it reads a bit like streaming thoughts :smile: 

Few things I will note:
* Yes, nightly inherently allows for things being broken or missing. There's some broken toolstates at the time of this post (you can always view tool status here: https://rust-lang.github.io/rustup-components-history/)
* You should consider whether you simply need _a_ nightly (in order to be able to utilize unstable features), or whether you always explicitly need _the latest_ nightly build
* There's various strategies folks employ depending on their respective answer to that question
    * If you just need _a_ nightly then a common approach is to utilize a [toolchain file](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) which is pinned to a known good nightly, and periodically updated
    * Note that rustup also has an `--allow-downgrade` flag that can be utilized to help with this too (e.g. `rustup toolchain install --allow-downgrade nightly -c rustfmt`) 
    * If you genuinely need _the latest_ nightly for compilation then you can consider things like multiple/matrix jobs (separate app compile/test from formatting), grabbing a precompiled binary, etc.

As for your questions:
> Any advice on how to get rustfmt working on Travis on nightly builds? Right now my builds succeed/fail intermittently depending on whether rustfmt is available, because I use cargo make and one of the steps is verifying the code is properly formatted (which I do want to keep as a test of the code).

Yes see above

> Any way to update my local environment to the latest nightly compiler and still have a working rustfmt? Right now when I develop locally I don't always get the latest compiler just because I happen to have rustfmt installed - and being able to run cargo fmt is extremely useful, not something I want to give up.

Yes see above. The toolchain file approach in particular can be helpful here, since that allows you to ensure your local env will match what happens in CI (both will have same toolchain, component versions, etc.)

> On its face, both issues would be resolved if there was a way of installing rustfmt by downloading a statically linked precompiled binary - the latest one, even if compiled with a slightly older rustc version - independently of the version of the currently installed rustc. If that means using curl instead of rustup, so be it. Is there such an alternative?

Yes, and I mentioned this in my previous response. We do provide precompiled versions for (most) of the Tier 1 platforms as artifacts attached to our [GitHub Releases](https://github.com/rust-lang/rustfmt/releases)
> one could also grab the binaries from our releases.

**Comment by calebcartwright** [comment]
> I guess what I'm asking for is a stable URL for a script install-latest-rustfmt.sh and possibly an equivalent uninstall-latest-rustfmt.sh which would revert things to their pristine state. This way a CI build could curl and execute this and then rustfmt would "just work" even if the current (nightly) toolchain can't compile rustfmt. Does that make sense?

Yes I understand what you'd like, but no, we're not going to do that for you. Fundamentally, you have to decide if you want the absolute bleeding edge nightly for both the compiler and formatter, and if so you have to accept that sometimes things will be broken or missing on that bleeding edge. You cannot have it both ways when it comes to stability and bleeding edge nightly.

There are a _lot_ of teams in the position of needing to build on nightly and also run formatting checks in their CI process, and there are a lot of viable, actively utilized approaches to support this, many of which have already been enumerated and discussed in this thread.
